Gradcheck small fixes

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54480 Implement faster gradcheck and enable for all ops except some
* #54479 Factor out numerical logic
* #54378 Refactor numerical to calculate wrt all outputs at once
* #54092 Refactor get numerical jacobian
* #54049 Refactor get analytical jacobian
* **#53916 Gradcheck small fixes**
* #53857 Refactor gradcheck

This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- ~add a test to see if when torch.allclose fails, we indeed return false.~
- ~uncomment test from previous PR.~

Differential Revision: [D27201692](https://our.internmc.facebook.com/intern/diff/D27201692)

soulitzer committed:
Fix gradcheck not returning false

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Fix gradcheck not returning false"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Fix gradcheck not returning false"




[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -591,37 +596,30 @@ def fn(input):
 
         if check_batched_grad:
-            test_batched_grad(fail_test, tupled_inputs, o, j)
+            test_batched_grad(fail_test, tupled_inputs, o, i)
```
fixes the output_idx (used only for the error message previously)

soulitzer commented on test/test_overrides.py:
```diff
@@ -804,7 +804,6 @@ def test_gradcheck(self):
         # Tensor-likes.
         self.assertEqual(total_used_attrs, {
             'data',
-            'device',
```
Previously, we are missing a return before `checkIfNumericalAnalyticAreClose`, now that we add this early exit, we may not access the device attr.

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -569,31 +575,24 @@ def fail_test(msg):
 
         if check_batched_grad:
-            test_batched_grad(fail_test, tupled_inputs, o, j)
+            return test_batched_grad(fail_test, tupled_inputs, o, i)
```
gradcheck should return False if test fails. 
Also j -> i, because we need the output idx. This helps fix the error message.

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Gradcheck small fixes"




[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -458,6 +459,11 @@ def _differentiable_outputs(x):
     return tuple(o for o in _as_tuple(x) if o.requires_grad)
 
 
+def get_notallclose_msg(a, n, i, j, error_str='') -> str:
```
nit: we can take this opportunity to improve these variable names, e.g.
a -> analytical
n -> numerical
i -> output_idx
j -> input_idx

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -562,31 +568,26 @@ def fn(input):
-                else:
-                    # R -> R, R -> C
-                    checkIfNumericalAnalyticAreClose(a, n, j)
+                if o.is_complex():    # C -> C, R -> C
+                    if not torch.allclose(analytical_from_imag_grad_out[j], numerical_from_imag_grad_out[j], rtol, atol):
+                        return fail_test(get_notallclose_msg(analytical_from_imag_grad_out[j],
+                                                             numerical_from_imag_grad_out[j], i, j,
+                                                             "Gradients failed to compare equal for grad output = 1j. "))
+                if inp.is_complex():  # C -> R, C -> C
+                    if not torch.allclose(a, n, atol, rtol):
+                        return fail_test(get_notallclose_msg(a, n, i, j,
+                                                             "Gradients failed to compare equal for grad output = 1. "))
+                else:                 # R -> R, R -> C
+                    if not torch.allclose(a, n, atol, rtol):
+                        return fail_test(get_notallclose_msg(a, n, i, j))
```
nit: I like the old checkIfNumericAnalyticAreClose better because it hides the atol and rtol details. Could we keep checkIfNumericalAnalyticAreClose but have it return a boolean?

soulitzer replied:
I think we'd still need to pass in rtol/atol unless we still have it nested. Wouldn't that defeat the purpose of trying to hide its details?

zou3519 replied:
Yes, the proposal was to have it nested because the original checkIfNumericAnalyticAreClose is nested inside the function

zou3519 reviewed (commented):
Changes LGTM, had some minor comments. Waiting on tests (that are listed in the TODO in the PR body) to approve.

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- uncomment test from previous PR.


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- add a test to see if when torch.allclose fails, we indeed return false.
- ~uncomment test from previous PR.~


[ghstack-poisoned]

soulitzer commented:
@zou3519 I've finished the todos in the PR body. 

Still not sure about `checkIfNumericalAnalyticAreClose` though. SInce I wanted to avoid having a function nested. 

I tried out a version where its not nested but I don't think its ideal either since we'd have to pass so many parameters.
```
def checkIfNumericAnalyticAreClose(fail_test, analytical, numerical, rtol, atol, output_idx, input_idx, error_str=''):
    if not torch.allclose(analytical, numerical, rtol, atol):
        return fail_test(get_notallclose_msg(analytical, numerical, output_idx, input_idx,
                                            "Gradients failed to compare equal for grad output = 1j. "))
    return True
```
```
for j, (a, n, inp) in enumerate(zip(analytical, numerical, inp_tensors)):
    if a.numel() != 0 or n.numel() != 0:
        if o.is_complex():    # C -> C, R -> C
            if not checkIfNumericAnalyticAreClose(fail_test, analytical_from_imag_grad_out[j], 
                                                numerical_from_imag_grad_out[j], 
                                                rtol, atol, i, j,
                                                "Gradients failed to compare equal for grad output = 1j. "):
                return False
        if inp.is_complex():  # C -> R, C -> C
            if not checkIfNumericAnalyticAreClose(fail_test, a, n, atol, rtol, i, j,
                                                "Gradients failed to compare equal for grad output = 1. "):
                return False
        else:                 # R -> R, R -> C
            if checkIfNumericAnalyticAreClose(a, n, atol, rtol, i, j):
                return fail_test(get_notallclose_msg(a, n, i, j))
```

zou3519 commented:
@soulitzer thanks for writing how it would look out. Let's keep it how it is in the PR then, I agree it looks nicer without a checkIfNumericAnalyticAreClose function.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4080,6 +4079,16 @@ def test_gradcheck_check_analytical_jacobian_attributes(self):
         # TODO: when grad_input is incorrect dtype/size
         pass
 
+    def test_gradcheck_jacobian_mismatch(self):
+        def fn(x):
+            y = x.clone()
+            y.register_hook(lambda x: x + 1e-2)
+            return y
+        x = torch.ones(2, 2, requires_grad=True)
+        with self.assertRaisesRegex(RuntimeError, 'Jacobian mismatch for output 0 with respect to input 0'):
+            gradcheck(fn, (x,))
+        self.assertFalse(gradcheck(fn, (x,), raise_exception=False))
+
```
There are a few gradcheck jacobian mismatch paths (this test is testing R->R functions), we should write tests to exercise all of them for completeness. It looks like checking a C->C function, C->R function, and R->C function would do the job

zou3519 reviewed (approved):
Some comments about the test, other than that, LGTM

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- ~add a test to see if when torch.allclose fails, we indeed return false.~
- ~uncomment test from previous PR.~


[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- ~add a test to see if when torch.allclose fails, we indeed return false.~
- ~uncomment test from previous PR.~

Differential Revision: [D27201692](https://our.internmc.facebook.com/intern/diff/D27201692)

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck small fixes"


This PR fixes some bugs that are made more clear by the previous refactor.
- make sure gradcheck returns false when its supposed to fail and when raise_exception=False.
- make sure when test_batched_grad fails, it returns false when raise_exception=False

Removing checkIfNumericalAnalyticAreClose made sense here to me because underneath its really doing `torch.allclose`, and using that directly instead of adding another opaque function to call seemed to make the code more clear.

TODO: 
- ~add a test to see if when torch.allclose fails, we indeed return false.~
- ~uncomment test from previous PR.~

Differential Revision: [D27201692](https://our.internmc.facebook.com/intern/diff/D27201692)

[ghstack-poisoned]
